pkg/controllers/management/authprovisioningv2/role.go

```diff
@@ -153,6 +158,78 @@ func (h *handler) OnChange(key string, rt *v3.RoleTemplate) (*v3.RoleTemplate, e
 	return rt, nil
 }
 
+func (h *handler) hasAnnotationsForDeletingCluster(annotations map[string]string, isClusterRole bool) (bool, error) {
+	clusterName, cOK := annotations[clusterNameLabel]
+	clusterNamespace, cnOK := annotations[clusterNamespaceLabel]
+	if !cOK || (!cnOK && !isClusterRole) {
+		// if this role doesn't have a clustername/namespace label it isn't one of the protected roles, so move on
+		return false, nil
+	}
+
+	mgmtCluster, mgmtErr := h.mgmtClusters.Get(clusterName)
+	if mgmtErr != nil && !apierror.IsNotFound(mgmtErr) {
+		return false, mgmtErr
+	}
+
+	cluster, err := h.clusters.Get(clusterNamespace, clusterName)
+	if err != nil && !apierror.IsNotFound(err) {
+		return false, err
+	}
+	if apierror.IsNotFound(mgmtErr) && apierror.IsNotFound(err) {
+		return false, nil
+	}
+	if cluster == nil && mgmtCluster == nil {
+		// cluster + err both being nil should not happen - but if it does, safe to delete the role
+		return false, nil
+	}
+	// if the err is nil, we can assume that the cluster is not nil
+	if (cluster != nil && cluster.DeletionTimestamp == nil) && (mgmtCluster != nil && mgmtCluster.DeletionTimestamp == nil) {
```

Comment:
I think you may need a separate case where one cluster is nil and the other cluster is not set for deletion.
```golang
	if (cluster != nil && cluster.DeletionTimestamp == nil ) && mgmtCluster == nil  {
```